### PR TITLE
[IGNITE-22121] Change parameters for disaster recovery partition states api

### PR DESCRIPTION
**Pull request by Phillippko**
https://issues.apache.org/jira/browse/IGNITE-22121

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/rest-api/src/main/java/org/apache/ignite/internal/rest/api/recovery/DisasterRecoveryApi.java [hunk 1]
```diff
@@ -41,34 +43,30 @@ public interface DisasterRecoveryApi {
...
-    @ApiResponse(responseCode = "200", description = "Partition states returned.")
-    @ApiResponse(responseCode = "500", description = "Internal error.",
-            content = @Content(mediaType = MediaType.PROBLEM_JSON, schema = @Schema(implementation = Problem.class)))
-    @ApiResponse(responseCode = "404", description = "Zone is not found.",
+    @ApiResponse(responseCode = "400", description = "Bad request.",
             content = @Content(mediaType = MediaType.PROBLEM_JSON, schema = @Schema(implementation = Problem.class)))
     @Produces(MediaType.APPLICATION_JSON)
-    CompletableFuture<LocalPartitionStatesResponse> getLocalPartitionStates(@PathVariable("zoneName") String zoneName);
+    CompletableFuture<LocalPartitionStatesResponse> getLocalPartitionStates(
+            @QueryValue
```
Comment (Phillippko) [inline modules/rest-api/src/main/java/org/apache/ignite/internal/rest/api/recovery/DisasterRecoveryApi.java:50]: I think, it should be possible to get rid of Optional here, but not sure how

File: modules/rest-api/src/main/java/org/apache/ignite/internal/rest/api/recovery/GlobalPartitionStatesResponse.java [hunk 2]
```diff
@@ -29,6 +30,7 @@
 @Schema(description = "Information about global partition states.")
 public class GlobalPartitionStatesResponse {
     @Schema
+    @JsonInclude
```
Comment (Phillippko) [inline modules/rest-api/src/main/java/org/apache/ignite/internal/rest/api/recovery/GlobalPartitionStatesResponse.java:33]: Without JsonInclude, if states are empty, resulting body is `{}`, so deserialization fails ("states is null"). But didn't see any other usages of JsonInclude, so it seems there is a better way to fix it

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/DisasterRecoveryManager.java [hunk 3]
```diff
@@ -193,65 +195,108 @@ public CompletableFuture<Void> resetPartitions(int zoneId, int tableId) {
...
 
-            CatalogZoneDescriptor zoneDescriptor = zoneDesciptorOptional.get();
-            zoneId = zoneDescriptor.id();
+            zoneIds = zoneDescriptors.stream()
+                    .map(CatalogObjectDescriptor::id)
+                    .collect(toSet());
         }
 
-        Set<NodeWithAttributes> logicalTopology = dzManager.logicalTopology();
+        Set<String> uppercaseNodeNames = nodeNames.stream()
```
Comment (tkalkirill) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/DisasterRecoveryManager.java:271]: Again, is it a rule that you capitalize node names?
Let's make the code related to searching for nodes a separate method.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/DisasterRecoveryManager.java [hunk 4]
```diff
@@ -193,65 +195,108 @@ public CompletableFuture<Void> resetPartitions(int zoneId, int tableId) {
...
+            zoneIds = Set.of();
         } else {
-            Optional<CatalogZoneDescriptor> zoneDesciptorOptional = catalog.zones().stream()
-                    .filter(catalogZoneDescriptor -> catalogZoneDescriptor.name().equals(zoneName))
-                    .findAny();
+            Set<String> uppercaseZoneNames = zoneNames.stream()
+                    .map(String::toUpperCase)
+                    .collect(toSet());
+
+            List<CatalogZoneDescriptor> zoneDescriptors = catalog.zones().stream()
```
Comment (tkalkirill) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/DisasterRecoveryManager.java:252]: In previous changes, zone names were compared for equality without converting to uppercase, why is this so now?

If you try the following code in `org.apache.ignite.internal.sql.engine.ItCreateTableDdlTest`, then two queries will be executed without errors, so I think converting to uppercase is not correct.

```
sql("CREATE ZONE TEST_ZONE WITH STORAGE_PROFILES='" + DEFAULT_STORAGE_PROFILE + "'");
sql("CREATE ZONE \"test_zone\" WITH STORAGE_PROFILES='" + DEFAULT_STORAGE_PROFILE + "'");
```

Let's make the code related to searching for nodes a separate method.

If I understand correctly, then the logic is missing; if the names of the zones are not specified, then we take all the zones.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/DisasterRecoveryManager.java [hunk 5]
```diff
@@ -193,65 +195,108 @@ public CompletableFuture<Void> resetPartitions(int zoneId, int tableId) {
...
+                    .collect(toSet());
         }
 
-        Set<NodeWithAttributes> logicalTopology = dzManager.logicalTopology();
+        Set<String> uppercaseNodeNames = nodeNames.stream()
+                .map(String::toUpperCase)
+                .collect(toSet());
+
+        Set<NodeWithAttributes> nodes = dzManager.logicalTopology().stream()
+                .filter(node -> nodeNames.isEmpty() || uppercaseNodeNames.contains(node.nodeName().toUpperCase()))
```
Comment (tkalkirill) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/DisasterRecoveryManager.java:276]: There is no need to filter if node names are missing.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/DisasterRecoveryManager.java [hunk 6]
```diff
@@ -193,65 +195,108 @@ public CompletableFuture<Void> resetPartitions(int zoneId, int tableId) {
...
+                .collect(toSet());
+
+        if (!nodeNames.isEmpty() && nodes.size() != uppercaseNodeNames.size()) {
```
Comment (tkalkirill) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/DisasterRecoveryManager.java:279]: Why not compare sets for equality, what if they coincide in size but not in content?

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/exceptions/NodesNotFoundException.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+import static org.apache.ignite.lang.ErrorGroups.DisasterRecovery.NODES_NOT_FOUND_ERR;
+
+import java.util.Set;
+import org.apache.ignite.internal.lang.IgniteInternalException;
+
+/** Exception is thrown when appropriate node can`t be found. */
+public class NodesNotFoundException extends IgniteInternalException {
+    private static final long serialVersionUID = -6295004626426857228L;
+
+    public NodesNotFoundException(Set<String> missingNodes) {
```
Comment (tkalkirill) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/exceptions/NodesNotFoundException.java:29]: missingNodes -> missingNodeNames

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/exceptions/PartitionsNotFoundException.java [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+import static org.apache.ignite.lang.ErrorGroups.DisasterRecovery.PARTITIONS_NOT_FOUND_ERR;
+
+import java.util.Set;
+import org.apache.ignite.internal.lang.IgniteInternalException;
+
+/** Exception is thrown when appropriate partition can`t be found. */
+public class PartitionsNotFoundException extends IgniteInternalException {
+    private static final long serialVersionUID = -9215416423159317425L;
+
+    public PartitionsNotFoundException(Set<Integer> missingPartitions) {
```
Comment (tkalkirill) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/exceptions/PartitionsNotFoundException.java:29]: missingPartitions -> missingPartitionIds

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/exceptions/ZonesNotFoundException.java [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+import static org.apache.ignite.lang.ErrorGroups.DistributionZones.ZONE_NOT_FOUND_ERR;
+
+import java.util.Set;
+import org.apache.ignite.internal.lang.IgniteInternalException;
+
+/** Exception is thrown when appropriate node can`t be found. */
+public class ZonesNotFoundException extends IgniteInternalException {
+    private static final long serialVersionUID = -8475588176132321568L;
+
+    public ZonesNotFoundException(Set<String> missingZones) {
```
Comment (tkalkirill) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/exceptions/ZonesNotFoundException.java:29]: missingZones -> missingZoneNames.

File: modules/rest/src/integrationTest/java/org/apache/ignite/internal/rest/recovery/ItDisasterRecoveryControllerTest.java [hunk 10]
```diff
@@ -18,139 +18,246 @@
...
+
     @Inject
     @Client(NODE_URL + "/management/v1/recovery/")
     HttpClient client;
 
-    @Override
-    protected int initialNodes() {
-        return 1;
+    @BeforeAll
+    public static void setUp() {
```
Comment (tkalkirill) [inline modules/rest/src/integrationTest/java/org/apache/ignite/internal/rest/recovery/ItDisasterRecoveryControllerTest.java:71]: I propose to add the case of zones with different cases in their names.

File: modules/rest-api/src/main/java/org/apache/ignite/internal/rest/exception/handler/IgniteInternalExceptionHandler.java [hunk 11]
```diff
@@ -37,9 +38,11 @@ public class IgniteInternalExceptionHandler implements ExceptionHandler<IgniteIn
 
     @Override
     public HttpResponse<? extends Problem> handle(HttpRequest request, IgniteInternalException exception) {
-        if (exception.code() == DistributionZones.ZONE_NOT_FOUND_ERR) {
+        if (exception.code() == DistributionZones.ZONE_NOT_FOUND_ERR
```
[the remark this answers is not in the document]
Comment (Phillippko) [reply modules/rest-api/src/main/java/org/apache/ignite/internal/rest/exception/handler/IgniteInternalExceptionHandler.java:41]: Error codes are not actually constant (result of `.registerErrorCode`) , can't be used in switch :(

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/DisasterRecoveryManager.java [hunk 12]
```diff
@@ -346,9 +407,16 @@ private void handleLocalPartitionStatesRequest(LocalPartitionStatesRequest reque
...
                     var tablePartitionId = (TablePartitionId) raftNodeId.groupId();
 
+                    if (!request.partitionIds().isEmpty() && !request.partitionIds().contains(tablePartitionId.partitionId())) {
+                        return;
+                    }
+
                     CatalogTableDescriptor tableDescriptor = catalogManager.table(tablePartitionId.tableId(), catalogVersion);
                     // Only tables that belong to a specific catalog version will be returned.
-                    if (tableDescriptor == null || request.zoneId() != NO_ZONE_ID && tableDescriptor.zoneId() != request.zoneId()) {
+                    if (tableDescriptor == null
```
Comment (tkalkirill) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/DisasterRecoveryManager.java:416]: I am a little confused by such a long condition, if possible you can separate it with additional parentheses, for example: `tableDescriptor == null || (!request.zoneIds().isEmpty() && !request.zoneIds().contains(tableDescriptor.zoneId()))`.

I would create a static method that would accept a collection and check the value like this: `!collection.isEmpty() && !collection.contains(e)`.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/DisasterRecoveryManager.java [hunk 13]
```diff
@@ -277,7 +280,65 @@ private CompletableFuture<Map<TablePartitionId, Map<String, LocalPartitionStateM
...
+                        .map(TablePartitionId::partitionId)
+                        .collect(toSet());
+
+                checkPartitions(foundPartitionIds, partitionIds);
+
+                return result;
+            });
+        }
+
+        return allOf(futures).handle((unused, throwable) -> result);
```
Comment (tkalkirill) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/DisasterRecoveryManager.java:295]: I wonder why you are swallowing an exception here in the branch on line 284, no, why?

File: modules/rest/src/integrationTest/java/org/apache/ignite/internal/rest/recovery/ItDisasterRecoveryControllerTest.java [hunk 14]
```diff
@@ -18,139 +18,292 @@
...
-public class ItDisasterRecoveryControllerTest extends ClusterPerTestIntegrationTest {
+public class ItDisasterRecoveryControllerTest extends ClusterPerClassIntegrationTest {
     private static final String NODE_URL = "http://localhost:" + Cluster.BASE_HTTP_PORT;
 
+    private static final Set<String> FILLED_ZONES = Set.of("first_ZONE", "second_ZONE", "third_ZONE");
+
+    private static final Set<String> MIXED_CASE_ZONES = Set.of("mixed_first_zone", "MIXED_FIRST_ZONE", "mixed_second_zone",
+            "MIXED_SECOND_ZONE");
+
+    private static final Set<String> ALL_ZONES = new HashSet<>(CollectionUtils.concat(FILLED_ZONES, MIXED_CASE_ZONES));
```
Comment (tkalkirill) [inline modules/rest/src/integrationTest/java/org/apache/ignite/internal/rest/recovery/ItDisasterRecoveryControllerTest.java:66]: Why without `EMPTY_ZONE`? =)

File: modules/rest/src/integrationTest/java/org/apache/ignite/internal/rest/recovery/ItDisasterRecoveryControllerTest.java [hunk 15]
```diff
@@ -18,139 +18,292 @@
...
     HttpClient client;
 
-    @Override
-    protected int initialNodes() {
-        return 1;
+    @BeforeAll
+    public static void setUp() {
+        ALL_ZONES.forEach(name -> {
+            sql(String.format("CREATE ZONE \"%s\" WITH storage_profiles='%s'", name, DEFAULT_AIPERSIST_PROFILE_NAME));
+            sql("CREATE TABLE \"" + name + "_table\" (id INT PRIMARY KEY, val INT) WITH PRIMARY_ZONE = '" + name + "'");
```
Comment (tkalkirill) [inline modules/rest/src/integrationTest/java/org/apache/ignite/internal/rest/recovery/ItDisasterRecoveryControllerTest.java:84]: Can also use `String#format`?

File: modules/rest/src/integrationTest/java/org/apache/ignite/internal/rest/recovery/ItDisasterRecoveryControllerTest.java [hunk 16]
```diff
@@ -18,139 +18,292 @@
...
+        });
+
+        sql("CREATE ZONE \"" + EMPTY_ZONE + "\" WITH storage_profiles='" + DEFAULT_AIPERSIST_PROFILE_NAME + "'");
```
Comment (tkalkirill) [inline modules/rest/src/integrationTest/java/org/apache/ignite/internal/rest/recovery/ItDisasterRecoveryControllerTest.java:87]: Can also use `String#format`?

File: modules/rest-api/src/main/java/org/apache/ignite/internal/rest/exception/handler/IgniteInternalExceptionHandler.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/rest-api/src/main/java/org/apache/ignite/internal/rest/exception/handler/IgniteInternalExceptionHandler.java:41]: But as far as I see `org.apache.ignite.lang.ErrorGroups.DistributionZones#ZONE_NOT_FOUND_ERR` as a numerical constant, why `switch` won’t work?

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/DisasterRecoveryManager.java (hunk 13 above)
Comment (Phillippko) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/DisasterRecoveryManager.java]: Fixed, now throwing it as DisasterRecoveryException (it was not handled before my changes, so created a new one). On line 284 I ignore the result of allOf (it is void)

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/DisasterRecoveryManager.java (hunk 12 above)
Comment (Phillippko) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/DisasterRecoveryManager.java]: Created a method containsOrEmpty, use it here and for nodes

File: modules/rest-api/src/main/java/org/apache/ignite/internal/rest/exception/handler/IgniteInternalExceptionHandler.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (Phillippko) [reply modules/rest-api/src/main/java/org/apache/ignite/internal/rest/exception/handler/IgniteInternalExceptionHandler.java]: Made a collection of bad request errors and check if this set contains the code

**Review by tkalkirill — APPROVED** [review]

**Pull request merged**